lib/grandpa/grandpa.go

```diff
@@ -449,188 +424,300 @@ func (s *Service) primaryBroadcastCommitMessage() {
 
-	// broadcast pre-vote
-	pv, err := s.determinePreVote()
-	if err != nil {
-		return err
+			switch action {
+			case determinePrevote:
+				isPrimary, err := s.handleIsPrimary()
+				if err != nil {
+					return err
+				}
+
+				// broadcast pre-vote
+				preVote, err := s.determinePreVote()
+				if err != nil {
+					return err
+				}
+
+				signedpreVote, prevoteMessage, err :=
+					s.createSignedVoteAndVoteMessage(preVote, prevote)
+				if err != nil {
+					return err
+				}
+
+				if !isPrimary {
+					s.prevotes.Store(s.publicKeyBytes(), signedpreVote)
+				}
+
+				s.sendPrevoteMessage(prevoteMessage)
+
+			case determinePrecommit:
+				preComit, err := s.determinePreCommit()
+				if err != nil {
+					return fmt.Errorf("determining precommit: %w", err)
+				}
+
+				signedpreComit, precommitMessage, err :=
+					s.createSignedVoteAndVoteMessage(preComit, precommit)
+				if err != nil {
+					return fmt.Errorf("creating signed vote: %w", err)
+				}
+
+				s.precommits.Store(s.publicKeyBytes(), signedpreComit)
+
+				s.sendPrecommitMessage(precommitMessage)
+
+			case finalize:
+				commitMessage, err := s.newCommitMessage(s.head, s.state.round, s.state.setID)
+				if err != nil {
+					return fmt.Errorf("generating commit message: %w", err)
+				}
+
+				commitConsensusMessage, err := commitMessage.ToConsensusMessage()
+				if err != nil {
+					return fmt.Errorf("transforming commit into consensus message: %w", err)
+				}
+
+				logger.Debugf("sending CommitMessage: %v", commitMessage)
+
+				s.network.GossipMessage(commitConsensusMessage)
+				s.telemetry.SendMessage(telemetry.NewAfgFinalizedBlocksUpTo(
+					s.head.Hash(),
+					fmt.Sprint(s.head.Number),
+				))
+
+				logger.Debugf("round completed in %s", time.Since(start))
+				return nil
+			}
+		}
 	}
+}
 
-	spv, vm, err := s.createSignedVoteAndVoteMessage(pv, prevote)
-	if err != nil {
-		return err
-	}
+// actions that should take place accordingly to votes the
+// finalisation engine knows about
+type action byte
 
-	if !isPrimary {
-		s.prevotes.Store(s.publicKeyBytes(), spv)
-	}
+const (
+	determinePrevote action = iota
+	determinePrecommit
+	finalize
+)
 
-	logger.Debugf("sending pre-vote message %s...", pv)
-	roundComplete := make(chan struct{})
-	// roundComplete is a signal channel which is closed when the round completes
-	// (will receive the default value of channel's type), so we don't need to
-	// explicitly send a value.
-	defer close(roundComplete)
+func (s *Service) finalizationEngine(cancel <-chan struct{}) <-chan action {
+	performAction := make(chan action)
 
-	// continue to send prevote messages until round is done
-	go s.sendVoteMessage(prevote, vm, roundComplete)
+	go func() {
+		defer close(performAction)
 
-	logger.Debug("receiving pre-commit messages...")
-	// through goroutine s.receiveVoteMessages(ctx)
-	time.Sleep(s.interval)
+		determinePrevoteTimer := time.NewTimer(s.interval * 2)
+		determinePrecommitTimer := time.NewTimer(s.interval * 4)
```

Comment:
we don't need to make these explicitly configurable?

Reply:
those timers are similar to what substrate has here -> https://github.com/paritytech/substrate/blob/c24431eb6a95197550b30715a4c124be1aea79de/client/finality-grandpa/src/environment.rs#L700